### PR TITLE
feat: base `CloudEvent` class as per v1 specs, including attribute validation

### PR DESCRIPTION
**Pull request by PlugaruT**
## Changes
The first PR towards a full rewrite of the library. What's included in this PR:
 - Support for v 1.0.2 of the specs with better validation errors and full support for custom extension attributes, including their validation. 
 - Full typing support
 - Better documentation for each class and methods

## One line description for the changelog

- [x] Tests pass
- [x] Appropriate changes to README are included in PR

### CODE REVIEW

File: src/cloudevents/core/v1/event.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+class CloudEvent:
+    def __init__(self, attributes: dict, data: Optional[dict] = None) -> None:
+        """
+        Create a new CloudEvent instance.
+
+        :param attributes: The attributes of the CloudEvent instance.
+        :type attributes: dict
+        :param data: The payload of the CloudEvent instance.
+        :type data: Optional[dict]
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:31]: there's no need to specify types here while we already have them in the method definition

File: src/cloudevents/core/v1/event.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+        :raises ValueError: If any of the required attributes are missing or have invalid values.
+        :raises TypeError: If any of the attributes have invalid types.
+        """
+        self._validate_attribute(attributes)
+        self._attributes = attributes
+        self._data = data
+
+    def _validate_attribute(self, attributes: dict) -> None:
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:40]: there's no need to have this method as an instance method. Let's convert to `static`. It'd also allow to test it separately and easier if needed

File: src/cloudevents/core/v1/event.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        """
+        Private method that validates the attributes of the CloudEvent as per the CloudEvents specification.
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:42]:
```suggestion
        Validates the attributes of the CloudEvent as per the CloudEvents specification.
        
        See https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#required-attributes
```

File: src/cloudevents/core/v1/event.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+            if not (5 <= len(custom_extension) <= 20):
+                raise ValueError(
+                    f"Extension attribute '{custom_extension}' should be between 5 and 20 characters long"
+                )
+
+            if not re.match(r"^[a-z0-9]+$", custom_extension):
+                raise ValueError(
+                    f"Extension attribute '{custom_extension}' should only contain lowercase letters and numbers"
+                )
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:119]: it's probably a good idea to:
1. split this method into multiple smaller ones with distinct validations.
2. maybe extract some reusable logic
3. maybe combine checks in order to avoid reporting attribute errors one-by-one. Meaning that we should go over all attributes and emit a combined error with all the possible attribute errors. Just as you did with required attributes

File: src/cloudevents/core/v1/event.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+    def get_attribute(self, attribute: str) -> Optional[Any]:
+        """
+        Retrieve a value of an attribute of the event denoted by the given `attribute`.
+
+        :param attribute: The name of the event attribute to retrieve the value for.
+        :type attribute: str
+
+        :return: The event attribute value.
+        :rtype: Optional[Any]
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:129]: no need for `:type` and `:rtype` while we're already using typed params and have return types

File: src/cloudevents/core/v1/event.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        if "dataschema" in attributes:
+            if not isinstance(attributes["dataschema"], str):
+                raise TypeError("Attribute 'dataschema' must be a string")
+
+            if not attributes["dataschema"]:
+                raise ValueError("Attribute 'dataschema' must not be empty")
+
+        for custom_extension in (
+            set(attributes.keys()) - REQUIRED_ATTRIBUTES - OPTIONAL_ATTRIBUTES
+        ):
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:100]: these are known in the spec as `extension attributes`

File: src/cloudevents/core/v1/event.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                raise ValueError(
+                    "Extension attribute 'data' is reserved and must not be used"
+                )
+
+            if not custom_extension[0].isalpha():
+                raise ValueError(
+                    f"Extension attribute '{custom_extension}' should start with a letter"
+                )
+
+            if not (5 <= len(custom_extension) <= 20):
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:111]: as per the [spec ](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#naming-conventions): `MUST be at least one character in length, and SHOULD NOT exceed 20 characters.`

File: src/cloudevents/core/v1/event.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+        for custom_extension in (
+            set(attributes.keys()) - REQUIRED_ATTRIBUTES - OPTIONAL_ATTRIBUTES
+        ):
+            if custom_extension == "data":
+                raise ValueError(
+                    "Extension attribute 'data' is reserved and must not be used"
+                )
+
+            if not custom_extension[0].isalpha():
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:106]: `SHOULD start with a letter.`, it's not a requirement. It's a recommendation.

File: tests/test_core/test_v1/test_event.py
Comment (xSAVIKx) [inline tests/test_core/test_v1/test_event.py:1]: we're missing positive cases

Comment (xSAVIKx) [reply tests/test_core/test_v1/test_event.py:1]: please make sure to add copyright notice to all the files. IDEs usually allow to configure a project-specific copyright that's applied automatically.

File: src/cloudevents/core/v1/event.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from typing import Any, Optional
+from datetime import datetime
+import re
+
+REQUIRED_ATTRIBUTES = {"id", "source", "type", "specversion"}
+OPTIONAL_ATTRIBUTES = {"datacontenttype", "dataschema", "subject", "time"}
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:20]:
```suggestion
REQUIRED_ATTRIBUTES: Final[set[str]] = {"id", "source", "type", "specversion"}
OPTIONAL_ATTRIBUTES: Final[set[str]]= {"datacontenttype", "dataschema", "subject", "time"}
```

File: src/cloudevents/core/v1/event.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        :type attributes: dict
+        :param data: The payload of the CloudEvent instance.
+        :type data: Optional[dict]
+
+        :raises ValueError: If any of the required attributes are missing or have invalid values.
+        :raises TypeError: If any of the attributes have invalid types.
+        """
+        self._validate_attribute(attributes)
+        self._attributes = attributes
+        self._data = data
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:38]: missing typings

File: src/cloudevents/core/v1/event.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+from typing import Any, Optional
+from datetime import datetime
+import re
+
+REQUIRED_ATTRIBUTES = {"id", "source", "type", "specversion"}
+OPTIONAL_ATTRIBUTES = {"datacontenttype", "dataschema", "subject", "time"}
+
+
+class CloudEvent:
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:23]: it's a good idea to expose at least `required` attributes with appropriate properties/methods on the class level.

**Review by xSAVIKx — COMMENTED** [review]
Hey @PlugaruT, thx for  PR. Please take a look at comments and suggestions.

File: src/cloudevents/core/v1/event.py (hunk 4 above)
Comment (PlugaruT) [reply src/cloudevents/core/v1/event.py:119]: My goal is to have this validation logic in a single place, it's easier to follow IMO and read it, rather than navigate into multiple small methods. 

About extracting reusable logic, I can try to look into this, but I also wanted to have it "match" the spec and treat each attribute as individual. Again, for readability reasons mainly. It's easier to reason about when the whole validation logic is colocated. 

> maybe combine checks in order to avoid reporting attribute errors one-by-one.

I'm a fan of "fail fast" approach, what would be an advantage of raising all exceptions at once?

File: src/cloudevents/core/v1/event.py (hunk 7 above)
Comment (PlugaruT) [reply src/cloudevents/core/v1/event.py:111]: Ah, don't know where I got the `5`. Good catch

File: src/cloudevents/core/v1/event.py (hunk 11 above)
Comment (PlugaruT) [reply src/cloudevents/core/v1/event.py:23]: There's the `get_attribute` method to get them by their name. I can replace it with dedicated getters for each attribute then, no probs

**Comment by PlugaruT** [comment]
There's a bit of a problem with `mypy` it seems, it doesn't like the `dict[str, Any]` for the `attributes` and also explicit `str` return for `id` field for example, since it's a required field and it's `str`. 
I tried to play with `TypedDict`, but that one doesn't allow for arbitrary keys, but we need to support custom extensions. So the only way to "fix" this is to suppress the error? Let me know what you think.

File: src/cloudevents/core/v1/event.py (hunk 4 above)
Comment (PlugaruT) [reply src/cloudevents/core/v1/event.py]: What do you think about something like this?
```
        CloudEvent._validate_required_attributes(attributes)
        CloudEvent._validate_attribute_types(attributes)
        CloudEvent._validate_optional_attributes(attributes)
        CloudEvent._validate_extension_attributes(attributes)
 ```
 and each method will hold the necesary validation logic?

Comment (xSAVIKx) [reply src/cloudevents/core/v1/event.py]: Grouping CloudEvents (CE) attributes validation with `required` and `optional` as well as having a separate `extension` rules is IMO a good idea.

On the fail fast approach - sure, that's a valid one, but IMO not in this case. I guess people would only see the validation errors when they are developing against CE, not during active usage (hopefully). So it makes sense to gather all the validation errors all together and present them together to ease analysis of a defective event. I'd probably even go and create a custom error instead of using standard ones with more details on what and where has failed the validation.

**Comment by xSAVIKx** [comment]
@PlugaruT, please don't use force push. The only merge strategy allowed in this repository is `Squash and merge`, so it's OK to have as many commits as you want and the way you want them.

force push makes it pretty hard to follow the review process in GitHub unfortunately

**Comment by PlugaruT** [comment]
> > There's a bit of a problem with `mypy` it seems, it doesn't like the `dict[str, Any]` for the `attributes` and also explicit `str` return for `id` field for example, since it's a required field and it's `str`. I tried to play with `TypedDict`, but that one doesn't allow for arbitrary keys, but we need to support custom extensions. So the only way to "fix" this is to suppress the error? Let me know what you think.
> 
> There's a `total` option for TypedDict. You can read more in MyPy docs: https://mypy.readthedocs.io/en/stable/typed_dict.html#mixing-required-and-non-required-items and in the PEP: https://peps.python.org/pep-0589/#totality
> 
> We can have `RequiredCloudEventAttributes(TypedDict)` and then have `CloudEventAttributes(RequiredCloudEventAttributes, total=False). This will also eliminate part of the validations. E.g. typings and presence of required attributes can be checked by the typing system and we can have only optional runtime validation. This may be a good approach especially where speed is more important than runtime validation.

Yes, but the problem is with arbitrary keys, like extension names, since the `TypedDict` doesn't accept random keys besides the ones defined. If I try to define a CloudEvent instance with an extension, mypy complains:
```
src/cloudevents/core/v1/event.py:252: error: Extra key "customextension" for
TypedDict "CloudEventsAttributes"  [typeddict-unknown-key]
        e = CloudEvent({
                       ^
Found 2 errors in 1 file (checked 1 source file)
```
There's a whole discussion going on here https ://github.com/python/mypy/issues/ 4617 (intended not to link the PR's). So yes, it will "work" for the known attributes, but users will have to set the flag when they'll use this. Up to you what we want to do here.

**Comment by PlugaruT** [comment]
@xSAVIKx, I addressed all the comments, except the mypy one that is a bit tricky. Let me know your thoughts

File: src/cloudevents/core/v1/exceptions.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+    def __init__(self, errors: dict[str, list[str]]) -> None:
+        super().__init__("Validation errors occurred")
+        self.errors: dict[str, list[str]] = errors
+
+    def __str__(self) -> str:
+        error_messages = [
+            f"{key}: {', '.join(value)}" for key, value in self.errors.items()
+        ]
+        return f"{super().__str__()}: {', '.join(error_messages)}"
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/exceptions.py:27]:
```suggestion
class CloudEventValidationError(Exception):
    """
    Holds validation errors aggregated during a CloudEvent creation.
    """

    def __init__(self, errors: dict[str, list[str]]) -> None:
        super().__init__("Validation errors occurred")
        self.errors: dict[str, list[str]] = errors
        """
        The errors gathered during the CloudEvent creation where key is the name of the attribute and
        value is a list of errors related to that attribute.
        """

    def __str__(self) -> str:
        error_messages = [
            f"{key}: {', '.join(value)}" for key, value in self.errors.items()
        ]
        return f"{super().__str__()}: {', '.join(error_messages)}"
```

File: src/cloudevents/core/v1/event.py [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+        attributes: dict[str, Any],
+    ) -> dict[str, list[str]]:
+        """
+        Validates the extension attributes.
+
+        :param attributes: The attributes of the CloudEvent instance.
+        :return: A dictionary of validation error messages.
+        """
+        errors = {}
+        extension_errors = []
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:161]:
```suggestion
        errors: dict[str, list[str]] = {}
        extension_errors: list[str] = []
```

File: src/cloudevents/core/v1/event.py [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+        self._data: Optional[dict] = data
+
+    @staticmethod
+    def _validate_attribute(attributes: dict[str, Any]) -> None:
+        """
+        Validates the attributes of the CloudEvent as per the CloudEvents specification.
+
+        See https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#required-attributes
+        """
+        errors = {}
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/event.py:60]:
```suggestion
        errors: dict[str, list[str]] = {}
```

File: src/cloudevents/core/v1/exceptions.py (hunk 12 above)
Comment (xSAVIKx) [reply src/cloudevents/core/v1/exceptions.py:27]: I've re-read the validation part. I suggest we change the validation to actually have `errors` where dict key is the attribute name and value holds all the errors related to that attribute.

**Review by xSAVIKx — COMMENTED** [review]
@PlugaruT overall LGTM. I'd suggest we modify the error gathering a bit, but otherwise that's good.

File: src/cloudevents/core/v1/exceptions.py (hunk 12 above)
Comment (PlugaruT) [reply src/cloudevents/core/v1/exceptions.py:27]: I've been thinking about this, but do you think this will be valuable, besides grouping the errors more logically, by the field 🤔 
I'm mostly thinking about programmatically reacting to these errors when using the library. Say I get a CloudEvent via HTTP, and I'm trying to use this library and I have some errors. I can handle them in code by checking the `required` or `type` or even `optional` fields and decide programmatically how to deal with that bad event, rather than having to check errors by substrings and based on fields. 
I'm 50/50 between these 2 approaches. Let me know.

Comment (xSAVIKx) [reply src/cloudevents/core/v1/exceptions.py:27]: If we're talking about some automated actions, then we can use actual typed errors instead of strings. That'd give the context of whether it's a type error or smth else + we keep the by attribute name grouping

Comment (PlugaruT) [reply src/cloudevents/core/v1/exceptions.py:27]: Let me see then, I think introducing the typed errors might be beneficial. Something like:
```
{
    "id": [TypeError("'id' field should be string"), MissingRequiredAttribute("'id'" attribute is required")]
    ...
```

File: src/cloudevents/core/v1/exceptions.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class MissingRequiredAttributeError(BaseCloudEventException):
+    """
+    Exception for missing required attribute.
+    """
+
+    def __init__(self, missing: str) -> None:
+        super().__init__(f"Missing required attribute: '{missing}'")
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/exceptions.py:44]:
```suggestion
    def __init__(self, attribute_name: str) -> None:
        super().__init__(f"Missing required attribute: '{attribute_name}'")
        self.attribute_name: str = attribute_name
```

File: src/cloudevents/core/v1/exceptions.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class CustomExtensionAttributeError(BaseCloudEventException):
+    """
+    Exception for invalid custom extension names.
+    """
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/exceptions.py:50]:
```suggestion
class CustomExtensionAttributeError(BaseCloudEventException):
    """
    Raised when a custom extension attribute violates naming conventions.
    """
```

File: src/cloudevents/core/v1/exceptions.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class InvalidAttributeTypeError(BaseCloudEventException):
+    """
+    Exception for invalid attribute type.
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/exceptions.py:55]:
```suggestion
    Raised when an attribute has an unsupported type. 
```

File: src/cloudevents/core/v1/exceptions.py [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+        super().__init__(f"Missing required attribute: '{missing}'")
+
+
+class CustomExtensionAttributeError(BaseCloudEventException):
+    """
+    Exception for invalid custom extension names.
+    """
+
+
+class InvalidAttributeTypeError(BaseCloudEventException):
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/exceptions.py:53]:
```suggestion
class InvalidAttributeTypeError(BaseCloudEventException, TypeError):
```

File: src/cloudevents/core/v1/exceptions.py [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+class MissingRequiredAttributeError(BaseCloudEventException):
+    """
+    Exception for missing required attribute.
+    """
+
+    def __init__(self, missing: str) -> None:
+        super().__init__(f"Missing required attribute: '{missing}'")
+
+
+class CustomExtensionAttributeError(BaseCloudEventException):
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/exceptions.py:47]:
```suggestion
class CustomExtensionAttributeError(BaseCloudEventException, ValueError):
```

File: src/cloudevents/core/v1/exceptions.py [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+        self.errors: dict[str, list[BaseCloudEventException]] = errors
+
+    def __str__(self) -> str:
+        error_messages = [
+            f"{key}: {', '.join(str(value))}" for key, value in self.errors.items()
+        ]
+        return f"{super().__str__()}: {', '.join(error_messages)}"
+
+
+class MissingRequiredAttributeError(BaseCloudEventException):
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/exceptions.py:38]:
```suggestion
class MissingRequiredAttributeError(BaseCloudEventException, ValueError):
```

File: src/cloudevents/core/v1/exceptions.py [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    Exception for missing required attribute.
```
Comment (xSAVIKx) [inline src/cloudevents/core/v1/exceptions.py:40]:
```suggestion
    Raised for attributes that are required to be present by the specification.
```

File: src/cloudevents/core/v1/exceptions.py (hunk 19 above)
Comment (xSAVIKx) [inline src/cloudevents/core/v1/exceptions.py:47]: I'd suggest we keep the attribute names as a parameter for this error.

File: src/cloudevents/core/v1/exceptions.py (hunk 18 above)
Comment (xSAVIKx) [inline src/cloudevents/core/v1/exceptions.py:53]: I'd suggest we keep the attribute names as a parameter for this error.

**Review by xSAVIKx — COMMENTED** [review]
@PlugaruT PTAL at the comments and suggestions. Doing good 👍

File: src/cloudevents/core/v1/event.py [hunk 22]
```diff
@@ -0,0 +1,299 @@
...
+        Validates the attributes of the CloudEvent as per the CloudEvents specification.
+
+        See https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#required-attributes
+        """
+        errors: dict[str, list] = defaultdict(list)
+        errors.update(CloudEvent._validate_required_attributes(attributes))
+        errors.update(CloudEvent._validate_optional_attributes(attributes))
+        errors.update(CloudEvent._validate_extension_attributes(attributes))
+        if errors:
+            raise CloudEventValidationError(dict(errors))
```
[the remark this answers is not in the document]
Comment (PlugaruT) [reply src/cloudevents/core/v1/event.py:73]: Just to cast it from `defaultdict` to `dict`. just my reflex :))

File: src/cloudevents/core/v1/event.py [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+
+        if "time" in attributes:
+            if not isinstance(attributes["time"], datetime):
+                errors["time"].append(InvalidAttributeTypeError("time", datetime))
+            if hasattr(attributes["time"], "tzinfo") and not attributes["time"].tzinfo:
+                errors["time"].append(
+                    InvalidAttributeValueError(
+                        "time", "Attribute 'time' must be timezone aware"
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (PlugaruT) [reply src/cloudevents/core/v1/event.py:142]: `time` attribute should respect RFC3339 which requires having timestamps with timezone attached.

[the remark this answers is not in the document]
Comment (PlugaruT) [reply src/cloudevents/core/v1/event.py]: I think here, we have 2 options, either default to UTC, but that would be a controversial behaviour, as we can't default the timezone of the clients, or, make it an explicit requirement. 
From my experience, using Python SDK and Java SDK, Java SDK will break when trying to have a naive datetime string. See [here](https://github.com/cloudevents/sdk-java/blob/main/core/src/main/java/io/cloudevents/core/v1/CloudEventV1.java#L88) for example, `time` is defined as `OffsetDateTime` which means exactly this, timestamp with a timezone offset

**Review by xSAVIKx — APPROVED** [review]

**Comment by PlugaruT** [comment]
@xSAVIKx can you also merge please,  I don't have the perms to do it. Thanks
